File: lib/data/models/recipes/recipe_model.dart
import 'package:calory_tool/data/models/recipes/recipe_nutrition_model.dart';
import 'package:hive_flutter/hive_flutter.dart';

part 'recipe_model.g.dart';

@HiveType(typeId: 10)
class RecipeModel extends HiveObject {
  RecipeModel({
    required this.ingredients,
    required this.types,
    this.id,
    this.name,
    this.description,
    this.image,
    this.nutrition,
  });

  factory RecipeModel.fromJson(Map<String, dynamic> json) {
    return RecipeModel(
      id: json['id'] as String?,
      name: json['name'] as String?,
      description: json['description'] as String?,
      image: json['image'] as String?,
      ingredients: List<String>.from(json['ingredients'] as List? ?? []),
      types: List<String>.from(json['types'] as List? ?? []),
      nutrition:
          json['nutrition'] != null
              ? RecipeNutritionModel.fromJson(
                json['nutrition'] as Map<String, dynamic>,
              )
              : null,
    );
  }

  @HiveField(0)
  final String? id;

  @HiveField(1)
  final String? name;

  @HiveField(2)
  final String? description;

  @HiveField(3)
  final String? image;

  @HiveField(4)
  final List<String> ingredients;

  @HiveField(5)
  final List<String> types;

  @HiveField(6)
  final RecipeNutritionModel? nutrition;
}

File: server/src/models/recipes/RecipeModel.js
import RecipeNutritionModel from "./RecipeNutritionModel.js";

class RecipeModel {
  constructor({ id, name, description, image, ingredients, types, nutrition }) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.image = image;
    this.ingredients = ingredients;
    this.types = types;
    this.nutrition = nutrition;
  }

  static fromJson(json) {
    return new RecipeModel({
      id: json["recipe_id"],
      name: json["recipe_name"],
      description: json["recipe_description"],
      image: json["recipe_image"],
      ingredients: json["recipe_ingredients"]?.["ingredient"] ?? [],
      types: json["recipe_types"]?.["recipe_type"] ?? [],
      nutrition: json["recipe_nutrition"] ? RecipeNutritionModel.fromJson(json["recipe_nutrition"]) : null,
    });
  }
}

export default RecipeModel;
